src/network/network_gaming_messages.cc

```diff
@@ -53,91 +88,36 @@ const std::string NetworkGamingMessages::get_message(const std::string& code,
 		   format("%s, %s, %s, %s", code, get_message(arg1), get_message(arg2), get_message(arg3)));
 	}
 
-	assert(ngmessages[code].n_arg <= 3);
+	const std::string msg_translated = _(ngmessages.at(code));
+	const size_t n_fmt_arg = format_impl::Tree::get(msg_translated).get_nodes_count();
```

Comment:
This should be delegated a bit more. Anything that's in an `impl` namespace is an implementation detail and should not be accessed directly. There should be one non-namespace function that provides the interface between the impl namespace and any code that wishes to use the functionality, like `::format()` which merely forwards its arguments to `format_impl::format()`. 

So please similarly add a function `unsigned ::count_placeholders(const std::string&)` or such to `base/string.cc`, which just calls another new function `unsigned format_impl::count_placeholders(const std::string&)` in `tree.h`, which in turn then queries the `Tree` object.